controller/controller_server/csi_controller_server.py

```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
             context.set_details('an internal exception occurred : {}'.format(ex))
             return csi_pb2.CreateVolumeResponse()
 
+    def _copy_volume_from_snapshot(self, vol, src_snapshot_id, min_vol_size, array_mediator):
+        vol_name = vol.volume_name
+        try:
+            src_snapshot = array_mediator.get_snapshot_by_id(src_snapshot_id)
+            if not src_snapshot:
+                raise controller_errors.SnapshotNotFoundError(src_snapshot_id)
+            src_snapshot_name = src_snapshot.snapshot_name
+            src_snapshot_capacity = src_snapshot.capacity_bytes
+            logger.debug("Copy snapshot {0} data to volume {1}.".format(src_snapshot_id, vol_name))
+            array_mediator.copy_volume_from_snapshot(vol_name, src_snapshot_name, src_snapshot_capacity,
+                                                     min_vol_size)
+            logger.debug("Copy Volume frm Snapshot finished")
+        except controller_errors.VolumeNotFoundError as ex:
+            logger.error("Volume not found while copying snapshot data to volume")
+            logger.exception(ex)
+            raise ex
+        except Exception as ex:
+            logger.error("Exception raised while copying snapshot data to volume")
+            logger.exception(ex)
+            try:
+                self._rollback_create_volume_from_snapshot(array_mediator, vol.id)
+            except Exception:
+                pass
+            raise ex
+        return vol
+
+    @retry(Exception, tries=5, delay=1)
```

Comment:
Lets open a new ticket to build up mechanism when NOT to do rollback.
Currently we will do rollback on each failure to avoid mistake in the white-listed exception

Reply:
CSI-1505

Reply:
I don't think this is about which exceptions cause rollback, I think it's about which exceptions cause retry.

we have always performed retires only on connection exception, and I don't see why it should be different here.

Reply:
leaving unresolved for future ticket

Reply:
CSI-1505